var/spack/repos/builtin/packages/mpich/package.py

```diff
@@ -93,7 +93,12 @@ class Mpich(AutotoolsPackage):
 '-Wl,-flat_namespace'.'''
     )
 
-    provides('mpi@:3.1')
+    variant('cuda', default=False, when="@3.4:", description="Enable CUDA support")
+    depends_on('cuda', when='+cuda')
+    conflicts('+cuda', when='device=ch3')
+
+    provides('mpi@:4.0')
+    provides('mpi@:3.1', when='@:4.0')
```

Comment:
```suggestion
    provides('mpi@:3.1', when='@:3.2')
```

If I understand correctly, that MPICH 3.2.x provides MPI-3.1.

Reply:
I see. So it says that MPICH provides MPI version < 3.1 when MPICH version is < 3.2?

Reply:
Hm, `@:3.1` means the open interval `(-inf, 3.2)` and the order on versions is lexicographical, so includes any `3.1.x` version too. So the change means: MPI <=3.1 is provided by mpich <= 3.2.anything. What we want is that MPI <= 3.1 is provided by >= 3.2, but apparently how it's expressed right now + spack's weird intersection behavior for "provides" as linked above make it correct...

Reply:
OK, I'm thoroughly confused right now. If you are saying your suggested change is correct, then I'll trust it and commit it to the PR.

Reply:
:laughing: it's correct